source/routing/journey_pattern_container.h

```diff
@@ -0,0 +1,191 @@
+/* Copyright © 2001-2015, Canal TP and/or its affiliates. All rights reserved.
+
+This file is part of Navitia,
+    the software to build cool stuff with public transport.
+
+Hope you'll enjoy and contribute to this project,
+    powered by Canal TP (www.canaltp.fr).
+Help us simplify mobility and open public transport:
+    a non ending quest to the responsive locomotion way of traveling!
+
+LICENCE: This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+Stay tuned using
+twitter @navitia
+IRC #navitia on freenode
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+
+#pragma once
+
+#include "raptor_utils.h"
+#include <boost/optional.hpp>
+
+namespace navitia { namespace type {
+
+struct PT_Data;
+struct DiscreteVehicleJourney;
+struct FrequencyVehicleJourney;
+struct StopTime;
+
+}}// namespace navitia::type
+
+namespace navitia { namespace routing {
+
+// TODO: constructor private with JourneyPatternContainer friend?
+struct JourneyPatternPoint {
+    JpIdx jp_idx;
+    SpIdx sp_idx;
+    uint16_t order;
+    bool operator==(const JourneyPatternPoint& other) const {
+        return sp_idx == other.sp_idx && order == other.order;
+    }
+};
+std::ostream& operator<<(std::ostream&, const JourneyPatternPoint&);
+
+
+// TODO: constructor private with JourneyPatternContainer friend?
+struct JourneyPattern {
+    std::vector<JppIdx> jpps;
+    std::vector<const type::DiscreteVehicleJourney*> discrete_vjs;
+    std::vector<const type::FrequencyVehicleJourney*> freq_vjs;
+    RouteIdx route_idx;
+    PhyModeIdx phy_mode_idx;
+
+    bool operator==(const JourneyPattern& other) const {
+        return jpps == other.jpps
+            && discrete_vjs == other.discrete_vjs
+            && freq_vjs == other.freq_vjs
+            && route_idx == other.route_idx;
+    }
+    template<typename VJ> std::vector<const VJ*>& get_vjs();
+    template<typename F> void for_each_vehicle_journey(const F f) const {
```

Comment:
Minor: I don't know if `const` is necessary, I mean, someone can define a functor and call `for_each_vehicle_journey()` to extract some info... For example, I wanted to extract all impacts that were applied on vj...